staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,287 @@
+	return &Manager{
+		client:           svClient,
+		apiserverID:      apiserverID,
+		crdInformer:      crdInformer,
+		svInformer:       svInformer,
+		queue:            workqueue.NewNamed("storageversion-updater"),
+		teardownCountMap: sync.Map{},
+	}
+}
+
+// UpdateActiveTeardownsCount updates the teardown count of the CRD in sync.Map.
+func (m *Manager) UpdateActiveTeardownsCount(crdName string, value int) error {
+	val, ok := m.teardownCountMap.Load(crdName)
+	if !ok {
+		return fmt.Errorf("error while trying to update number of teardowns. No entry found in sync.Map for %v", crdName)
+	}
+	newTeardownCount := val.(int) + value
+	m.teardownCountMap.Store(crdName, newTeardownCount)
+	return nil
+}
+
+// Enqueue adds the CRD name to the SV upadte queue.
+func (m *Manager) Enqueue(crdName string) {
+	if crdName == "" {
+		return
+	}
+	klog.V(4).Infof("enqueueing crd %s for storageversion update", crdName)
+	m.queue.Add(crdName)
+}
+
+// DeleteTeardownCountInfo deletes specified key from the sync.Map.
+func (m *Manager) DeleteTeardownCountInfo(crdName string) {
+	m.teardownCountMap.Delete(crdName)
+}
+
+// Sync runs a goroutine over the SV updater queue to indefinitely
+// process any queued storageversion updates unless stopCh is invoked.
+func (m *Manager) Sync(stopCh <-chan struct{}, workers int) {
+	defer m.queue.ShutDownWithDrain()
+	klog.V(2).Infof("starting storageversion sync loop")
+	for i := 0; i < workers; i++ {
+		go wait.Until(func() {
+			m.worker()
+		}, workerFrequency, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (m *Manager) worker() {
+	for m.processLatestUpdateFor() {
+	}
+}
+
+func (m *Manager) processLatestUpdateFor() bool {
+	ctx := context.TODO()
+	key, quit := m.queue.Get()
+	defer m.queue.Done(key)
+	if quit {
+		return false
+	}
+
+	// Fetch the latest CRD from CRD informer cache to update the storageversion for it.
+	crdName := key.(string)
+	crd, err := m.crdInformer.Lister().Get(crdName)
+
+	if err != nil {
+		// No need to update storgaversion if CRD not found.
+		if errors.IsNotFound(err) {
+			klog.V(4).Infof("crd: %s not found in CRD cache, abandoning storageversion update", crdName)
+			return true
+		}
+
+		klog.V(4).Infof("error fetching crd: %s from CRD cache, will retry storageversion update", crdName)
+		// TODO: enqueue again?
+		return true
+	}
+
+	skip, reason, err := m.shouldSkipSVUpdate(crd)
+	if err != nil {
+		klog.V(4).Infof("error while checking if storagversion update can be skipped for crd: %s, err: %v", crd.Name, err)
+		// TODO: enqueue again?
+		return true
+	}
+
+	if skip {
+		klog.V(4).Infof("skipping storagversion update for crd: %s, reason: %s ", crd.Name, reason)
+		return true
+	}
+
+	klog.V(4).Infof("starting storageversion update for crd: %s", crd.Name)
+	err = m.updateStorageVersion(ctx, crd)
+	if err == nil {
+		klog.V(4).Infof("successfully updated storage version for %s", crd.Name)
+		return true
+	}
+
+	// TODO: indefinitely requeue on error?
+	m.Enqueue(crd.Name)
```

Comment:
This smells like a hot loop.  What prevents this from immediately de-queuing on the `m.queue.Get()`?   Perhaps structuring with the separate `syncHandler` like our example in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md#rough-structure will help.